### ISSUE TITLE
Alternative download source

### ISSUE THREAD

**Issue by IzzySoft** [SamAmco/track-and-graph#2]
Would you mind providing the APK at an alternative place (e.g. attaching it to the Github releases) – for the folks outside of Google Play? Some of us have no access, or prefer to keep Google off their devices. Us "tin foils" especially prefer tracker-free apps like yours (just [checked it with Exodus](https://reports.exodus-privacy.eu.org/en/reports/104492/): congrats!).

Thanks in advance!

**Comment by SamAmco** [comment]
Good idea I will try to sort this out soon.

**Comment by IzzySoft** [comment]
Great, thanks in advance! Please announce here when it's there, so folks can subscribe to this issue to get notified.

**Comment by SamAmco** [comment]
@IzzySoft You can now find the latest release apk uploaded under releases: https://github.com/SamAmco/track-and-graph/releases/tag/v1.3.3

I will upload apk's there for future releases also.

**Comment by IzzySoft** [comment]
Great, thanks a lot! To spread the word, I've just added it to my repo (should turn up [here](https://apt.izzysoft.de/fdroid/index/apk/com.samco.trackandgraph) with the next sync in about 20 hours) and hope that's fine with you. If you agree, your app could also be included with the [official F-Droid repo](https://f-droid.org/), where it would be built from source. That place has more "prestige" than my 3rd party repo and is more trusted. It's home of thousands of FOSS apps (and FOSS only) – your app would perfectly fit there. In case you've not yet heard of F-Droid and want a quick overview: [F-Droid: The privacy-friendly alternative to Google Play Store](https://android.izzysoft.de/articles/named/fdroid-intro-1) is an article (series) I wrote about it.

(Disclosure: I'm one of the F-Droid maintainers)

**Comment by SamAmco** [comment]
My pleasure. Thanks for spreading the word, I very much appreciate it. F-Droid is on my todo list for Track & Graph, I've had a few people suggest that now. I've never uploaded anything there before so hopefully it will be easy :)

**Comment by IzzySoft** [comment]
Oh, you don't upload anything there – we pull it :smile: Two things we'd need from your end:

* to automatically detect updates, our checker compares tag names with `versionCode`/`versionName` of the build gradle – but it only RegEx-parses the latter, so [variables, functions or calculations](https://github.com/SamAmco/track-and-graph/blob/v1.3.3/app/build.gradle#L27) won't work. Would it be possible to have "literal values" instead? Alternatively, they could also be in a separate file consisting of just two lines (`versionCode` and `versionName`), so we can check against that.
* optional, but highly recommended: could you establish [Fastlane file structure](https://gitlab.com/snippets/1895688) in your repo to provide screenshots (and be able to maintain summary+description yourself without the need for MRs on every change)? You can use the very same for Play Store deployment, too.

I've just created a [Request For Packaging](https://gitlab.com/fdroid/rfp/issues/1187) (RFP) you might wish to follow, and will update it accordingly with the above information when available. Thanks a lot!

**Comment by SamAmco** [comment]
Ok I've updated the build gradle to use hardcoded version name and code and I will look into fastlane soon. Can you confirm for me what tags will be built and published by the build server. Ideally it would only build tags without -rc extension (so 1.3.3 but not 1.4.1-rc until it is tagged with just 1.4.1)?

Also I can see you RFP has raised security issues. Is this something I need to fix?

**Comment by IzzySoft** [comment]
Thanks for your help, Samuel! I cannot confirm something will be built before we tried it :laughing: – but yes, that's the idea.

As for which tags are "pulled": that indeed can be controlled via tag names. There's one corresponding field in F-Droid metadata: `AutoUpdateMode: Version v%v` (would mean: take only `v<versionName>`, ignore all else) – so you could tag RCs just e.g. "rc-versionName" or simply skip the leading "v", and they'd be ignored by us. Of course we can adjust this pattern if you prefer something different (e.g. `AutoUpdateMode: Version release-v%v` – in which case the releases you wish to be picked up should be tagged `release-v<versionName>`).

Some projects even go a step further and manually open merge requests for "not-stable" builds – just adding the new version without increasing `CurrentVersion`/`CurrentVersionCode`. The effect is those versions would not be "suggested" to the user (not even as update) – but the user could decide to manually install them from within the client.

As for the security issues: None of the two affect F-Droid builds, so both are rather hints for your own builds. You see e.g. `INFO: Removing gradle-wrapper.jar at gradle/wrapper/gradle-wrapper.jar` (we replace that with our own – remember, no binaries). The second one you can fix as described – again, for your own builds.

**Comment by SamAmco** [comment]
Great, thank you for helping me set me up on F-Droid it's very much appreciated. I will just stick with v%v and use rc-v%v for beta builds from now on. Hopefully it all works smoothly, I will keep an eye on the RFP. Other than that, let me know if you need anything else from me! :)

**Comment by IzzySoft** [comment]
Sounds good! Let me know when the first tag with Fastlane is there (we need that tagged or it won't be used), so I can update metadata in the RFP and "push it upstairs".

**Comment by SamAmco** [comment]
@IzzySoft  just published a new release to production tagged 1.4.3

**Comment by IzzySoft** [comment]
Thanks! I've updated metadata accordingly and marked the RFP for our packagers. Not sure if they need `en-US` in fastlane (as that's the usual fallback) – but then they'll say so.

If there's no further activity in the RFP for 2+ weeks, be welcome to ping me again so I can check where it might hang.

**Comment by SamAmco** [comment]
Alright thanks for all your help :)

**Comment by SamAmco** [comment]
@IzzySoft I think the RFP might be stuck?

**Comment by IzzySoft** [comment]
Missed in lots of work I guess. I've pinged our packagers to get it out of the bin. Should nothing happen for another 2 weeks, feel free to ping me again.

**Comment by IzzySoft** [comment]
@SamAmco unstuck! RFP now has been processed, your app should show up with the next index, which is hopefully published within a few days from now. Congrats!

PS: When it appears on f-droid.org, I'll remove it again from my repo after a short delay. No need to have it "duplicated" – and we also better avoid the confusion caused by "signature mismatch" (two different signatures: yours in my repo, F-Droid's there; cross-updates are not possible).

**Comment by SamAmco** [comment]
@IzzySoft Awesome, thanks I will keep an eye out for it on F-Droid!

**Comment by SamAmco** [comment]
@all-contributors please add @IzzySoft  for infra

**Comment by IzzySoft** [comment]
Ui… thanks! :star_struck:

And a cool bot you have there :smiley:

**Comment by SamAmco** [comment]
Hi @IzzySoft I'm told the last couple of releases have not been picked up by F-Droid? I'm not sure the best way to go about fixing this. Since you set it up before I don't really have a handle on how F-Droid works. Would you mind taking a look, or pointing me in the right direction?

**Comment by IzzySoft** [comment]
Often my snippets are good starting points. For all around builds & Co, it would be [this one](https://gitlab.com/snippets/1896503). Let's check for [your app's last build](https://f-droid.org/wiki/index.php?title=com.samco.trackandgraph/lastbuild&redirect=no) …

(half an hour later) waiting for that page to load, no idea what's so slow there now … OK, let's go [directly to the last build](https://f-droid.org/wiki/page/com.samco.trackandgraph/lastbuild_10600). Yes, that works. And it says:

```
Git checkout of 'v1.6.0' failed
detail begin

error: pathspec 'v1.6.0' did not match any file(s) known to git
detail end
```

Well, you added another prefix to that tag (`rc-`), so it should have been skipped. That run was on 2020-05-23 16:00:13Z, no new one since. And the one before succeeded as you say – which means we still have to wait for the next run.

Wait, according to metadata the last one we have is 1.5.1. Checking the tag v1.5.2, `build.gradle` identifies the app as "1.5.1", so it didn't match the tag name and hence was skipped. Checking v1.6.2: this should work again, tag name matches versionName at least.

**TL;DR:** 2 mismatches on your end, newer ones were not yet fetched.

PS: Thanks for asking, learned a new thing. Obviously the updater doesn't go for the tag first, else it wouldn't have tried to fetch v1.6.0. It must have seen "1.6.0" in the `build.gradle` and looked for a matching tag. Funny.

**Comment by SamAmco** [comment]
Excellent, thanks for the help here. I will follow this process next time :)
